core/services/ocrcommon/adapters.go

```diff
@@ -71,3 +80,110 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
 func (c *OCR3ContractTransmitterAdapter) FromAccount() (ocrtypes.Account, error) {
 	return c.ct.FromAccount()
 }
+
+var _ ocr3types.OnchainKeyring[[]byte] = (*OCR3OnchainKeyringMultiChainAdapter)(nil)
+
+type OnchainSigningStrategy interface {
+	ConfigCopy() job.JSONConfig
+	KeyBundleID(name string) (string, error)
+}
+
+type OCR3OnchainKeyringMultiChainAdapter struct {
+	keyBundles map[string]ocr2key.KeyBundle
+	publicKey  ocrtypes.OnchainPublicKey
+	lggr       logger.Logger
+}
+
+func NewOCR3OnchainKeyringMultiChainAdapter(ks keystore.OCR2, st OnchainSigningStrategy, lggr logger.Logger) (*OCR3OnchainKeyringMultiChainAdapter, error) {
+	keyBundles := map[string]ocr2key.KeyBundle{}
+	for name := range st.ConfigCopy() {
+		kbID, err := st.KeyBundleID(name)
+		if err != nil {
+			return nil, err
+		}
+		kb, err := ks.Get(kbID)
+		if err != nil {
+			return nil, err
+		}
+		keyBundles[name] = kb
+	}
+	// We don't need to check for the existence of `publicKey` in the keyBundles map because it is required on validation on `validate/validate.go`
+	return &OCR3OnchainKeyringMultiChainAdapter{keyBundles, keyBundles["publicKey"].PublicKey(), lggr}, nil
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) PublicKey() ocrtypes.OnchainPublicKey {
+	return a.publicKey
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) Sign(digest ocrtypes.ConfigDigest, seqNr uint64, r ocr3types.ReportWithInfo[[]byte]) (signature []byte, err error) {
+	info := new(structpb.Struct)
```

Comment:
@vyzaldysanchez You have a little bit of duplicated code here to handle unmarshaling the info and extracting the value you need. Can we extract this out to a private getter?